From 42c1964bdc2311ec142453cf43a8737da81b2f0d Mon Sep 17 00:00:00 2001
From: Jim Meyering <meyering@redhat.com>
Date: Wed, 12 Sep 2012 10:43:17 +0200
Subject: [PATCH 25/89] tests: show that small dev now evokes "mklabel gpt"
 diagnostic

* tests/t0203-gpt-create-on-min-sized-device.sh: New test.
* tests/Makefile.am (TESTS): Add it.
---
 tests/Makefile.am                             |  1 +
 tests/t0203-gpt-create-on-min-sized-device.sh | 47 +++++++++++++++++++++++++++
 2 files changed, 48 insertions(+)
 create mode 100644 tests/t0203-gpt-create-on-min-sized-device.sh

diff --git a/tests/Makefile.am b/tests/Makefile.am
index c97400b..66b9361 100644
--- a/tests/Makefile.am
+++ b/tests/Makefile.am
@@ -16,6 +16,7 @@ TESTS = \
   t0202-gpt-pmbr.sh \
   t0203-gpt-tiny-device-abort.sh \
   t0203-gpt-shortened-device-primary-valid.sh \
+  t0203-gpt-create-on-min-sized-device.sh \
   t0205-gpt-list-clobbers-pmbr.sh \
   t0206-gpt-print-with-corrupt-primary-clobbers-pmbr.sh \
   t0207-IEC-binary-notation.sh \
diff --git a/tests/t0203-gpt-create-on-min-sized-device.sh b/tests/t0203-gpt-create-on-min-sized-device.sh
new file mode 100644
index 0000000..113e191
--- /dev/null
+++ b/tests/t0203-gpt-create-on-min-sized-device.sh
@@ -0,0 +1,47 @@
+#!/bin/sh
+# parted 3.1 and prior would exit with no diagnostic when failing
+# to create a GPT partition table on a device that was too small.
+
+# Copyright (C) 2012 Free Software Foundation, Inc.
+
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 3 of the License, or
+# (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+. "${srcdir=.}/init.sh"; path_prepend_ ../parted
+require_512_byte_sector_size_
+
+dev=loop-file
+ss=$sector_size_
+
+# Create the smallest file that can accommodate a GPT partition table.
+dd if=/dev/null of=$dev bs=$ss seek=67 || framework_failure
+
+# create a GPT partition table
+parted -s $dev mklabel gpt > out 2>&1 || fail=1
+# expect no output
+compare /dev/null out || fail=1
+
+# Create a file that is 1 sector smaller, and require failure,
+# *with* a diagnostic.
+rm -f $dev
+dd if=/dev/null of=$dev bs=$ss seek=66 || framework_failure
+
+echo Error: device is so small it cannot even accommodate GPT headers \
+  > exp || framework_failure
+
+# Try to create a GPT partition table in too little space.  This must fail.
+parted -s $dev mklabel gpt > out 2>&1 && fail=1
+# There must be a diagnostic.
+compare out exp || fail=1
+
+Exit $fail
-- 
1.8.5.3

